nixos/modules/installer/tools/nixos-prepare-root.sh

```diff
@@ -0,0 +1,126 @@
+# We copy the Nix closure first because we need to invoke its tools  to register store path validity
+for i in $nix_closure $other_closures; do
+    # We support closures both in the format produced by `nix-store --export` and by `exportReferencesGraph`,
+    # mostly because there doesn't seem to be a single format that can be produced outside of a nix build and
+    # inside one. See https://github.com/NixOS/nix/issues/1242 for more discussion.
+    if [[ "$i" =~ \.closure$ ]]; then
+        echo "importing serialized closure $i to $mountPoint..."
+        fakechroot -- chroot $mountPoint nix-store --import < $i
+    else
+        echo "copying closure $i to $mountPoint..."
+        for j in $(perl @pathsFromGraph@ $i); do
+            echo -n "  $j... "
+	    # FIXME: this might actually be kind of expensive, might not be worth it... should probably
+	    # just track the root of each closure and check if that's valid, and if so skip the entire
+	    # set of paths it brings along with it, like the old NixOS install did for the Nix closure
+	    if true; then # ! fakechroot -- chroot $mountPoint nix-store --check-validity $j 2> /dev/null; then
+                rsync -a $j $mountPoint/nix/store/
+	        echo "copied"
+	    else
+                echo "already valid"
+            fi
+        done
+        # Register the paths in the input closures as valid. This is necessary
+        # to prevent them from being deleted the first time we install
+        # something.  (I.e., Nix will see that, e.g., the glibc path is not
+        # valid, delete it to get it out of the way, but as a result nothing
+        # will work anymore.)
+
+        fakechroot -- chroot $mountPoint nix-store --register-validity < $i
+    fi
+done
+
+# Create the required /bin/sh symlink; otherwise lots of things
+# (notably the system() function) won't work.
+mkdir -m 0755 -p $mountPoint/bin
+# !!! assuming that @shell@ is in the closure
+ln -sf @shell@ $mountPoint/bin/sh
+
+echo "setting the system closure to '$system'..."
+fakechroot -- chroot $mountPoint nix-env \
+    "${extraBuildFlags[@]}" -p /nix/var/nix/profiles/system --set "$system"
+
+ln -sfn /nix/var/nix/profiles/system $mountPoint/run/current-system
+
+# Copy the NixOS/Nixpkgs sources to the target as the initial contents
+# of the NixOS channel.
+mkdir -m 0755 -p $mountPoint/nix/var/nix/profiles
+mkdir -m 1777 -p $mountPoint/nix/var/nix/profiles/per-user
+mkdir -m 0755 -p $mountPoint/nix/var/nix/profiles/per-user/root
+# srcs=$(nix-env "${extraBuildFlags[@]}" -p /nix/var/nix/profiles/per-user/root/channels -q nixos --no-name --out-path 2>/dev/null || echo -n "")
+if [ -z "$noChannelCopy" ] && [ -n "$channel" ]; then
+    echo "copying channel..."
+    fakechroot -- chroot $mountPoint nix-env --option build-use-substitutes false \
+        "${extraBuildFlags[@]}" -p /nix/var/nix/profiles/per-user/root/channels --set "$channel" --quiet
```

Comment:
There is a `channel_closure` a little farther down, but you're right, I probably screwed something up here.